src/Mvc/Mvc.Core/src/Formatters/TranscodingWriteStream.cs

```diff
@@ -0,0 +1,156 @@
+            await _stream.FlushAsync(cancellationToken);
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException();
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException();
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            ThrowArgumentException(buffer, offset, count);
+            var bufferSegment = new ArraySegment<byte>(buffer, offset, count);
+            return WriteAsync(bufferSegment, flush: false, cancellationToken);
+        }
+
+        private async Task WriteAsync(
+            ArraySegment<byte> bufferSegment,
+            bool flush,
+            CancellationToken cancellationToken)
+        {
+            var decoderCompleted = false;
+            while (!decoderCompleted)
+            {
+                _decoder.Convert(
+                    bufferSegment,
+                    _charBuffer.AsSpan(_charsDecoded),
+                    flush,
+                    out var bytesDecoded,
+                    out var charsDecoded,
+                    out decoderCompleted);
+
+                _charsDecoded += charsDecoded;
+                bufferSegment = bufferSegment.Slice(bytesDecoded);
+
+                if (flush || !decoderCompleted)
+                {
+                    // This is being invoked from FlushAsync or the char buffer is not large enough
+                    // to accomodate all writes.
+                    await WriteBufferAsync(flush, cancellationToken);
+                }
+            }
+        }
+
+        private async Task WriteBufferAsync(bool flush, CancellationToken cancellationToken)
+        {
+            var encoderCompletd = false;
+            var charsWritten = 0;
+            var byteBuffer = ArrayPool<byte>.Shared.Rent(_maxByteBufferSize);
+
+            try
+            {
+                while (!encoderCompletd && charsWritten < _charsDecoded)
+                {
+                    _encoder.Convert(
+                        _charBuffer.AsSpan(charsWritten, _charsDecoded - charsWritten),
+                        byteBuffer,
+                        flush,
+                        out var charsEncoded,
+                        out var bytesUsed,
+                        out encoderCompletd);
+
+                    await _stream.WriteAsync(byteBuffer.AsMemory(0, bytesUsed), cancellationToken);
+                    charsWritten += charsEncoded;
+                }
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(byteBuffer);
```

Comment:
Clear the array before returning?

Reply:
We generally don't clear these buffers e.g. https://github.com/aspnet/AspNetCore/blob/master/src/Http/WebUtilities/src/HttpResponseStreamWriter.cs#L78

Reply:
For some context, see recent corefx discussion on clearing:
https://github.com/dotnet/corefx/pull/35927#discussion_r264083647

tl;dr

We don't have consistent guidance on when to clear atm, but could be helpful for defense in depth.